plugins/GranularPitchShifter/GranularPitchShifterEffect.h

```diff
@@ -0,0 +1,168 @@
+/*
+ * GranularPitchShifter.h
+ *
+ * Copyright (c) 2024 Lost Robot <r94231/at/gmail/dot/com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#ifndef LMMS_GRANULAR_PITCH_SHIFTER_EFFECT_H
+#define LMMS_GRANULAR_PITCH_SHIFTER_EFFECT_H
+
+#include "Effect.h"
+#include "GranularPitchShifterControls.h"
+
+#include "BasicFilters.h"
+#include "interpolation.h"
+
+namespace lmms
+{
+
+constexpr float PrefilterBandwidth = 0.96f;// 96% of nyquist
+constexpr double GlideSnagRadius = 0.001;
+constexpr int SafetyLatency = 3;
+constexpr float RangeSeconds[5] = {5, 10, 40, 40, 120};
+
+
+class GranularPitchShifterEffect : public Effect
+{
+public:
+	GranularPitchShifterEffect(Model* parent, const Descriptor::SubPluginFeatures::Key* key);
+	~GranularPitchShifterEffect() override = default;
+	bool processAudioBuffer(sampleFrame* buf, const fpp_t frames) override;
+
+	EffectControls* controls() override
+	{
+		return &m_granularpitchshifterControls;
+	}
+	
+	// double index and fraction are required for good quality
+	float getHermiteSample(double index, int ch) {
+		const int index_floor = static_cast<int>(index);
+		const double fraction = index - index_floor;
+		
+		float v0, v1, v2, v3;
+
+		if (index_floor == 0) {
+			v0 = m_ringBuf[m_ringBuf.size() - 1][ch];
+		} else {
+			v0 = m_ringBuf[index_floor - 1][ch];
+		}
```

Comment:
Another K&R snuck in I'm sorry 😭but we can at least do this
```suggestion
		if (index_floor == 0) { v0 = m_ringBuf[m_ringBuf.size() - 1][ch]; }
		else { v0 = m_ringBuf[index_floor - 1][ch]; }
```